File: src/bcd_api/schemas/hold.py
"""Pydantic schemas for Hold model."""

from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from src.bcd_api.schemas.common import TimestampMixin
from src.shared.constants import HoldStatus


class HoldCreate(BaseModel):
    """Schema for creating a new hold."""

    borrower_id: int = Field(..., description="Borrower database ID")
    bibliographic_record_id: int = Field(..., description="Bibliographic record ID")
    created_by: Optional[str] = Field(None, max_length=100, description="Librarian who placed hold")
    notes: Optional[str] = Field(None, description="Additional notes")

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "borrower_id": 1,
                "bibliographic_record_id": 1,
                "created_by": "librarian@school.fr",
            }
        }
    )


class HoldResponse(TimestampMixin):
    """Schema for hold response."""

    id: int
    borrower_id: int
    bibliographic_record_id: int
    hold_date: datetime
    queue_position: int
    status: HoldStatus
    available_date: Optional[datetime]
    expiration_date: Optional[date]
    fulfilled_date: Optional[datetime]
    notified: bool
    notification_method: Optional[str]
    created_by: Optional[str]
    notes: Optional[str]

    model_config = ConfigDict(from_attributes=True)


class HoldSummary(BaseModel):
    """Summary schema for hold (for lists)."""

    id: int
    borrower_id: int
    borrower_name: str
    bibliographic_record_id: int
    title: str
    queue_position: int
    status: HoldStatus
    hold_date: datetime

    model_config = ConfigDict(from_attributes=True)


class HoldReadyForPickup(BaseModel):
    """Schema for holds ready for pickup."""

    id: int
    borrower_id: int
    borrower_name: str
    class_name: Optional[str]
    title: str
    available_date: datetime
    expiration_date: date
    expires_in_days: int

    model_config = ConfigDict(from_attributes=True)


class HoldWithDetails(HoldResponse):
    """Hold response with borrower and bibliographic record details."""

    borrower_name: Optional[str] = None
    borrower_string_id: Optional[str] = None
    borrower_class: Optional[str] = None
    title: Optional[str] = None
    authors: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)

File: src/bcd_web_vue/js/models/hold.js
/**
 * Hold (Réservation) data models with TypeScript JSDoc type definitions.
 * Fully aligned with Pydantic and SQLAlchemy models.
 */

/**
 * Basic hold response details
 * @typedef {Object} Hold
 * @property {number} id - Hold database ID
 * @property {number} borrower_id - Borrower database ID
 * @property {number} bibliographic_record_id - Bibliographic record database ID
 * @property {string} hold_date - Date and time placed (ISO 8601)
 * @property {number} queue_position - Position in the queue
 * @property {string} status - Hold status: 'pending', 'ready_for_pickup', 'fulfilled', 'cancelled', 'expired'
 * @property {string|null} available_date - Date made available for pickup (ISO 8601)
 * @property {string|null} expiration_date - Date when pickup hold expires (YYYY-MM-DD)
 * @property {string|null} fulfilled_date - Date when hold was fulfilled (ISO 8601)
 * @property {boolean} notified - Whether notification has been sent
 * @property {string|null} notification_method - Notification method
 * @property {string|null} created_by - User/Librarian who placed the hold
 * @property {string|null} notes - Additional notes
 */

/**
 * Hold summary used in lists
 * @typedef {Object} HoldSummary
 * @property {number} id - Hold database ID
 * @property {number} borrower_id - Borrower database ID
 * @property {string} borrower_name - Borrower full name
 * @property {number} bibliographic_record_id - Bibliographic record database ID
 * @property {string} title - Book title
 * @property {number} queue_position - Position in hold queue
 * @property {string} status - Hold status
 * @property {string} hold_date - Hold date and time (ISO 8601)
 */

/**
 * Hold detailed info with complete borrower and record details
 * @typedef {Object} HoldWithDetails
 * @property {number} id - Hold database ID
 * @property {number} borrower_id - Borrower database ID
 * @property {number} bibliographic_record_id - Bibliographic record database ID
 * @property {string} hold_date - Date and time placed (ISO 8601)
 * @property {number} queue_position - Position in the queue
 * @property {string} status - Hold status
 * @property {string|null} available_date - Date made available for pickup (ISO 8601)
 * @property {string|null} expiration_date - Date when pickup hold expires (YYYY-MM-DD)
 * @property {string|null} fulfilled_date - Date when hold was fulfilled (ISO 8601)
 * @property {boolean} notified - Whether notification has been sent
 * @property {string|null} notification_method - Notification method
 * @property {string|null} created_by - User/Librarian who placed the hold
 * @property {string|null} notes - Additional notes
 * @property {string|null} borrower_name - Borrower full name
 * @property {string|null} borrower_string_id - Borrower student ID
 * @property {string|null} borrower_class - Borrower class name
 * @property {string|null} title - Bibliographic record title
 * @property {string|null} authors - Bibliographic record author(s)
 */

// Export empty object to make this a module
export {};
